listenbrainz/spark/background.py

```diff
@@ -0,0 +1,200 @@
+
+
+class BackgroundJobProcessor:
+    """ To handle long-running tasks without skipping heartbeats in the main thread, a background message processor
+    that runs in a separate thread.
+
+    This setup is useful when processing a message (even rarely) can take over a minute. Processing such tasks
+    in the main thread can cause missed heartbeats which in turn lead to the connection being restarted and the same
+    messages being redelivered to the consumer setting up a loop. The rabbitmq queue essentially gets stuck and the
+    consumer keeps processing the same messages again and again.
+
+    This processor runs in its own thread. However, that presents another challenge because messages need to be ack-ed
+    from the same thread on which they were received (the restriction is about channels ack-ing from the same
+    channel but channels are not thread-safe, hence same thread). Here's a brief explanation of how the overall flow
+    works:
+
+        1. The RabbitMQ consumer listens for messages on the main thread.
+        2. Upon receiving a message, the RabbitMQ callback enqueues the message into the processor's message queue.
+        3. The processor handles the message and, upon completion, enqueues the message into its acknowledgment queue.
+        4. The main thread periodically checks the processor's acknowledgment queue for completed messages and acknowledges them.
+    """
+
+    def __init__(self, app):
+        self.app = app
+        self.thread = None
+        self.done = False
+        self.internal_message_queue = Queue()
+        self.internal_message_ack_queue = Queue()
+
+        self.response_handlers = {}
+        self.register_handlers()
+
+    def terminate(self):
+        """ Stop the background job processor and its thread """
+        self.done = True
+        self.thread.join()
+
+    def enqueue(self, message):
+        """ Add a message for processing to internal queue """
+        self.internal_message_queue.put(message, block=False)
+
+    def pending_acks(self):
+        """ Add a processed message to internal queue for acknowledging the message to rabbitmq. """
+        messages = []
+        while True:
+            try:
+                message = self.internal_message_ack_queue.get(block=False)
+                messages.append(message)
+            except Empty:
+                break
+        return messages
+
+    def run(self):
+        """ Infinite loop that keeps processing messages enqueued in the internal message queue and puts them on
+         ack queue if successfully processed.
+        """
+        with self.app.app_context():
+            while not self.done:
+                try:
+                    message = self.internal_message_queue.get(block=True, timeout=5)
+                    self.process_message(message)
+                    self.internal_message_ack_queue.put(message)
+                except Empty:
+                    self.app.logger.debug("Empty internal message queue")
```

Comment:
You probably don't want this in production long term...

Reply:
Yeah, that's why I set it to debug log level and we log by default at INFO level. Its only for debugging in future if something hangs.